Add Validating/MutatingWebhookConfiguration structured Apply

Danil-Grigorev opened the pull request:
This PR adds `Apply` method in the library for `ValidatingWebhookConfigurations` and `MutatingWebhookConfigurations`, to efficiently update existing resources, and preserve the functionality of `service-ca-bundle` `inject-cabundle` annotations.

The functionality relies solely on generation changes done by the server. To track generation changes and prevent unnecessary polling on resource updates, the helper generation caching is introduced. It should be used similarly to:
```golang
func (optr *Operator) syncValidatingWebhook() error {
	expectedGeneration := resourcemerge.ExpectedValidatingWebhooksConfiguration(mapiv1.NewValidatingWebhookConfiguration().Name, optr.generations)
	validatingWebhook, updated, err := resourceapply.ApplyValidatingWebhookConfiguration(optr.kubeClient.AdmissionregistrationV1(),
		events.NewLoggingEventRecorder(optr.name),
		mapiv1.NewValidatingWebhookConfiguration(), expectedGeneration)
	if err != nil {
		return err
	}
	if updated {
		resourcemerge.SetValidatingWebhooksConfigurationGeneration(&optr.generations, validatingWebhook)
	}

	return nil
}
```

This PR is tested with e2e [coverage](https://github.com/openshift/cluster-api-actuator-pkg/blob/7f3bb702a1baababc269747d2b463240686ce7b6/pkg/operators/machine-api-operator.go#L75-L193) in MAO with https://github.com/openshift/machine-api-operator/pull/642

Fixes https://github.com/openshift/library-go/issues/953

Danil-Grigorev commented:
/retest

mfojtik commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,152 @@
+		}
+	}
+
+	// Explcitily specify type for requiredUnstr to get object meta accessor
```
nit: typo

sttts commented on pkg/operator/resource/resourceapply/unstructured.go:
```diff
@@ -0,0 +1,91 @@
+// usntructured will be merged with the existing usntructured and an update performed if the
+// usntructured spec and metadata differ from the previously required spec and metadata.
+// For further detail, check the top-level comment.
+func applyUnstructured(resourceClient dynamic.ResourceInterface, path string, recorder events.Recorder,
```
doesn't this ask for hot loops? Apply in a generic way does not really make sense IMO.

Danil-Grigorev replied:
So you think this needs an option to `forceApply`, same as it is for Deployments and DaemonSets? Or it should repeat the attempt couple of times before failing?

JoelSpeed replied:
@Danil-Grigorev Can you please build this into the MAO and try it on a real cluster to see how it behaves?

I don't think it should hot loop, it's based on the `ApplyDeployment` logic, so if that doesn't hot loop I think this shouldn't? If the metadata and spec that we input doesn't change, then the spec hash annotation shouldn't change and we should short circuit on line 42 before we attempt to update again. @sttts Is there something in this in particular that makes you think it would hot loop?

sttts replied:
This func looks generic. But it is wrong in most cases. It does not consider things like list-type markers (compare https://github.com/kubernetes/enhancements/blob/6a1ac61fbfd64d49e13fae3a8d773dd4ebda53c8/keps/sig-api-machinery/0006-apply.md). It might be fine for one specific resource. But having this helper here encourages simple, but wrong code.

Danil-Grigorev replied:
Once this PR passes e2e tests, we would have the proof that hot loops do not occur in current implementation: https://github.com/openshift/machine-api-operator/pull/642

sttts replied:
@Danil-Grigorev please check https://github.com/openshift/library-go/pull/836#issuecomment-658822346. This is a blocker for the PR.

sttts commented on pkg/operator/resource/resourceapply/unstructured.go:
```diff
@@ -0,0 +1,91 @@
+
+// This is based on resourcemerge.EnsureObjectMeta but uses the metav1.Object interface instead
+// TODO: Update this to use resourcemerge.EnsureObjectMeta or update resourcemerge.EnsureObjectMeta to use the interface
+func ensureObjectMeta(existing, required metav1.Object) (bool, error) {
```
what do we do for other resources? I looks wrong to have this here in a generic way, but not use it for other resources as well.

JoelSpeed replied:
Other resources use https://github.com/openshift/library-go/blob/100bf3ff5a1911717545c1f77513bdb5270ddea6/pkg/operator/resource/resourcemerge/object_merger.go#L10-L17 which does the same thing but doesn't use the `metav1.Object` interface, even though it could.

We could update the public implementation so we only have one version but that would be a breaking change to the method since it accepts a `*metav1.ObjectMeta` and a `metav1.ObjectMeta`, the latter not satisfying the interface

sttts commented:
Why is this using unstructured after all an not a typed client as we do for all other resources?

/hold

JoelSpeed commented:
> Why is this using unstructured after all an not a typed client as we do for all other resources?

For background on why we are proposing this change: 
We have, during the 4.6 cycle, introduced webhooks into the Machine API project. Currently, these are being applied by CVO, with a create only annotation so that the CA Bundle can be injected without conflicting with the CVO. 

We would like to be able to reconfigure the MWC/VWC resources in the future so looked at how we could do this. For now, we looked at bringing the management of the MWC/VWC into the MAO with our other resources. Other resources being the Deployment and DaemonSet we manage. They use the library-go `resourceapply.ApplyDeployment` and `resourceapply.ApplyDaemonset` helpers to apply so that we can be a bit more clever about the changes.

We then thought it would be good to have the same for MWC and VWC. However, the logic for `ApplyDeployment` and `ApplyDaemonset` is pretty much duplicated, the algorithm is the same bar some initial setup and the types that are fed in.
We thought it might be good to make a reusable implementation of the apply algorithm so that we can more easily test the logic, and to make it reusable, it has to be unstructured. The idea being that all of the `ApplyMyThing` methods could do their initial set up (eg copying the CABundle across if needed for MWC/VWC), then use the shared, underlying `Apply` algorithm that is unstructured.

The algorithm we have in `applyUnstructured` matches that in `ApplyDeployment` and `ApplyDaemonset`, we hash the spec that we are given, put that in an annotation, check that the metadata hasn't changed (ie the spec hash hasn't changed), check that the generation hasn't been changed since we last updated, and then only update if one of these things has changed., overwriting the entire spec with the given input. We could replace the internal logic of `ApplyDeployment` and `ApplyDaemonset` in theory with this implementation without breaking anything.

By copying the CABundles specifically in the MWC/VWC implementation, we can get around the problem that an external party needs to modify an otherwise managed object, but because we do that before the spec hash is calculated, we are safe for future updates.

There is prior art for using unstructured in this package as `ApplyServiceMonitor` uses unstructured, though they have a slightly different implementation that merges the spec here, perhaps we should look to use that instead? https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/monitoring.go#L24-L48

While I agree that ideally we would use server side apply, I didn't think that was ready for general consumption yet? This logic replicates what we have already with other resources that have a spec, if the spec has changed, we overwrite it completely. CVO does this today for many resources already, and this just mimics that, so I don't think we are making things worse really?

Long term, it is my belief that CVO should be able to manage MWC and VWC resources and handle the update to the CABundle (at the moment it uses a generic apply which just replaces the spec verabatim if there's a difference), if there is a better way to approach this problem and solve the issue for the wider Openshift community then I'm happy to help implement it

JoelSpeed commented:
FWIW, I've manually tested this code which is being used in https://github.com/openshift/machine-api-operator/pull/642 this morning and it is working pretty much as expected, there was one minor discrepancy but that was within the MAO repo and I believe it has now been fixed. I did not observe any hot loops, as I modified the resource spec it got put back and kept the CABundle as desired.

JoelSpeed commented:
Having said all of that, having a generic implementation that relies on updating the spec doesn't help here (this was originally implemented for a different type), since webhooks don't have a spec (so not actually sure why it appeared to be working). I think it would be best to have a typed version for these resources after all, since they aren't fitting the generic format for resources having a spec anyway

Danil-Grigorev commented:
> Having said all of that, having a generic implementation that relies on updating the spec doesn't help here (this was originally implemented for a different type), since webhooks don't have a spec (so not actually sure why it appeared to be working). I think it would be best to have a typed version for these resources after all, since they aren't fitting the generic format for resources having a spec anyway

Can't really agree with that, although I see advantages for typed approach. But just a note - current implementation requires to provide a path to resource spec equivalent, which is used to calculate hash. This should be as generic as it could be, with this line in place - https://github.com/openshift/library-go/pull/836/files#diff-61be38f5023c7c6da801d664793070d5R25

JoelSpeed commented:
> Can't really agree with that, although I see advantages for typed approach. But just a note - current implementation requires to provide a path to resource spec equivalent, which is used to calculate hash. This should be as generic as it could be, with this line in place - https://github.com/openshift/library-go/pull/836/files#diff-61be38f5023c7c6da801d664793070d5R25

I was not on the ball this morning, ignore that previous comment, it was completely wrong 🤦  Too much context switching right now

Danil-Grigorev commented:
/retest

sttts commented:
@JoelSpeed I understand your motiviation to add webhook support here. And I agree it makes sense. But for this very resourceapply package it is harmful to have a generic implementation. All generic implementations are wrong if defaulting happens in the apiserver. This even applies to specific implementations if upstream adds defaults and we forget to update our apply funcs. Hence, we have to be super careful with the logic here for each single resource type, otherwise we end up in hot loops.

tl/dr: Not having too many tempting generic helpers is a feature of this package, not code smell.

Maybe we can have some common logic like for ObjectMeta if that makes sense. Am not against that.

JoelSpeed commented:
> All generic implementations are wrong if defaulting happens in the apiserver. This even applies to specific implementations if upstream adds defaults and we forget to update our apply funcs. Hence, we have to be super careful with the logic here for each single resource type, otherwise we end up in hot loops.

Ok, this makes sense. How would you suggest we approach this then? Should we rewrite this to be structured and specific for the MWC/VWC, would specific implementations for these be acceptable for this package? This would then allow us to add CVO support for MWC/VWC, do you think that would be accepted?  

Or should we just keep our own implementation and manage the webhooks from MAO long term?

> Maybe we can have some common logic like for ObjectMeta if that makes sense. Am not against that.

How do you mean for that? I don't follow.

sttts commented:
>> Maybe we can have some common logic like for ObjectMeta if that makes sense. Am not against that.
> How do you mean for that? I don't follow.

You had some logic about ObjectMeta which looked generic. If it is and  matches what we do everywhere, that could make sense. But this should go to another independent PR.

sttts commented:
> Should we rewrite this to be structured and specific for the MWC/VWC, would specific implementations for these be acceptable for this package?

Yes. Similarly as for the other types. Be careful with the defaults that kube applies. You can find the defaulters here:

  https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/admissionregistration/v1/defaults.go

You have to manually replicate that logic. And yes, this is super tedious and error-prone. That's why we must be so careful in this package. Manual apply requires this kind of deep knowledge. This is the reason for [server side apply](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/0006-apply.md) upstream that we can eventually use some day.

JoelSpeed commented:
Ok, thanks for the guidance @sttts! I will work with @Danil-Grigorev over the next few days to re-approach this

JoelSpeed commented on pkg/operator/resource/resourcemerge/admissionregistration.go:
```diff
@@ -0,0 +1,76 @@
+	EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+
+	// we stomp everything
+	if !equality.Semantic.DeepEqual(existing.Webhooks, required.Webhooks) {
+		*modified = true
+		existing.Webhooks = required.Webhooks
+	}
```
Is there any defaulting within the `Webhooks` that we need to account for here? We could end up in a hot loop if there's server side defaulting that we are missing

JoelSpeed commented on pkg/operator/resource/resourcemerge/admissionregistration.go:
```diff
@@ -0,0 +1,76 @@
+	EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+
+	// we stomp everything
+	if !equality.Semantic.DeepEqual(existing.Webhooks, required.Webhooks) {
+		*modified = true
+		existing.Webhooks = required.Webhooks
+	}
```
Is there any defaulting within the Webhooks that we need to account for here? We could end up in a hot loop if there's server side defaulting that we are missing

Danil-Grigorev replied:
Having to call defaulting in here is the reason for possible hot loops. If we pull server implementation for webhooks, someone would have to suffer a burden of maintaining it in sync with upstream, or otherwise, any defaulting implementation change would be a 100% cause for hot looping. The location @sttts pointed at, is an internal `kubenetes/kubernetes` package. If it is not exposed at https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/api/admissionregistration/v1, I don't think it is wise to use it and especially, copying that. I'm not sure that the strict rules for kubernetes api changes are applied there, so the probability of described situation is higher. Let's leave it to the server-side apply, as a future improvement.

Right now I could agree that if you immediately apply the updated resource, returned from `ApplyMutatingWebhookConfiguration` or `ApplyValidatingWebhookConfiguration`once again, it would not be filtered by the logic, and `Update` call will still happen. Even if the resource is technically the same, from the algorithmic perspective you provided a different input, and the change should occur.

I'm adding the missing comment from `resourceapply/apps.go` from where the implementation was inspired, so the explanation will be preserved: https://github.com/openshift/library-go/blob/3b4f06f8efeda380cd78d17b09ed6d72edc1b06a/pkg/operator/resource/resourceapply/apps.go#L21-L51

JoelSpeed replied:
While I agree with that comment for the deployment implementation, this does not match with the logic that you have in this implementation.

We are checking the `Webhooks` for equality here, we aren't solely relying on the spec hash. There is no way that the two sets of webhooks will be equal unless the input has already been defaulted, so this will currently cause a hot loop. 

We have two options, we copy the logic from the K8s package (definitely do not import it), or we remove the lines I've highlighted here and solely rely on the spec hash annotation for detecting differences.

> someone would have to suffer a burden of maintaining it in sync with upstream, or otherwise, any defaulting implementation change would be a 100% cause for hot looping.

This has already been discussed, this is what Stefan was saying about in his previous comments. We have to be careful when adding to this package because if we don't get the defaulting right, we will hot loop. He's already said it's annoying because we have to keep them up to date. See https://github.com/openshift/library-go/pull/836#issuecomment-659425034

Danil-Grigorev replied:
Now I see it, thanks. Updated.

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,185 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// The Apply<type> methods in this file ensure that a resource is created or updated to match
```
this big block of go doc must go to doc.go. This describes generic behaviour that must not differ by resource. Obviously, this poses the question whether we have a generic behaviour. If not, we have to make it uniform.

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,185 @@
+		return nil, false, err
+	}
+
+	existing, err := client.Get(context.TODO(), required.GetName(), metav1.GetOptions{})
```
add context as first arg of the apply func

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,185 @@
+		}
+	}
+
+	toWrite := existing.DeepCopy()
```
this comes from a client. You own it. No need for deepcopy.

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,185 @@
+	modified := pointer.BoolPtr(false)
+	resourcemerge.EnsureObjectMeta(modified, &toWrite.ObjectMeta, required.ObjectMeta)
+	// there was no change to metadata, the generation was right
+	if !*modified && toWrite.ObjectMeta.Generation == expectedGeneration {
```
do these webhook support generation at all?

Danil-Grigorev replied:
Yes, they do, and the way it would be used should looks similar to this: https://github.com/openshift/machine-api-operator/blob/c9a1f7584b4bd025e1e99444d0f6281c86a18a1a/pkg/operator/sync.go#L169-L177

Unfortunately, the fake client does not support generation changes, so to test this we have e2e tests for our operator (https://github.com/openshift/cluster-api-actuator-pkg/pull/176/files#diff-56cc6cb7098e6f7d33d5d2812f6a1aa3R102-R127) and more specific ones: https://github.com/openshift/machine-api-operator/blob/c9a1f7584b4bd025e1e99444d0f6281c86a18a1a/pkg/operator/sync_test.go#L313-L316, both of which are using the real client.

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,185 @@
+
+func copyMutatingWebhookCABundle(required *admissionregistrationv1.MutatingWebhookConfiguration,
+	existing admissionregistrationv1.MutatingWebhookConfiguration) error {
+	existingMutatingWebhooks := make(map[string]admissionregistrationv1.MutatingWebhook)
```
set capacity to that of existing

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,185 @@
+	return actual, true, err
+}
+
+func copyMutatingWebhookCABundle(required *admissionregistrationv1.MutatingWebhookConfiguration,
```
s/required/from/
s/existing/to/

Isn't copy(to, from) more common in Go?

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,185 @@
+		}
+		webhooks[i] = mutatingWebhook
+	}
+	required.Webhooks = webhooks
```
note that Webhooks here is the same type as for validating webhook. You can easily merge this func with copyValidatingWebhookCABundle by using another function type.

sttts commented on pkg/operator/resource/resourcemerge/admissionregistration.go:
```diff
@@ -0,0 +1,51 @@
+		Name:           actual.Name,
+		LastGeneration: actual.ObjectMeta.Generation,
+	})
+}
```
these functions here irritate me. Where will they be used?

The ones for deployments come from the fact that deployments often are the primary resources deployed by an operator. But webhook probably are not.

Danil-Grigorev replied:
Just updated the PR description with links and a code sample showing how it is supposed to be used. If the user does not require generation tracking and could tolerate multiple update calls to the server, just set the `expectedGeneration` to 0 like here: https://github.com/openshift/library-go/pull/836/files#diff-895b3186a55d12e7a602b6c33d4881042ac6f3aad924a271d3fabb3693236b0aR173

or use the generic apply method.

Danil-Grigorev commented:
/hold cancel
/remove-lifecycle stale

Danil-Grigorev commented:
/remove-lifecycle rotten

jsafrane commented on pkg/operator/resource/resourceapply/generic.go:
```diff
@@ -170,12 +170,12 @@ func ApplyDirectly(clients *ClientHolder, recorder events.Recorder, manifests As
 				result.Error = fmt.Errorf("missing kubeClient")
 			}
-			result.Result, result.Changed, result.Error = ApplyValidatingWebhookConfiguration(clients.kubeClient.AdmissionregistrationV1(), recorder, t)
+			result.Result, result.Changed, result.Error = ApplyValidatingWebhookConfiguration(clients.kubeClient.AdmissionregistrationV1(), recorder, t, 0)
```
Since they need expectedGeneration now, should we remove them from `ApplyDirectly`? It will keep overwriting the existing object and that's not the behavior we want to support.

Danil-Grigorev replied:
This behavior is equal to other `ApplyDirectly` methods https://github.com/openshift/library-go/blob/c8d0c55677d43fef621e8afb5a2c092ed75a5baf/pkg%2Foperator%2Fresource%2Fresourceapply%2Fstorage.go#L90 I added a test case which shows how it will work.

jsafrane replied:
`CSIDriver` does not have `Generation` and can't be (easily) compared, direct apply is the only way. Anyway, I am not going to use `ApplyDirectly` / StaticController, so feel free to ignore this comment.

jsafrane reviewed (commented):
I tested `ApplyValidatingWebhookConfiguration` and it works as expected. I got one Generation bumped by caBundle injector, but after that things are pretty stable.

BTW, most (all?) my comments apply interchangeably both to `ApplyMutatingWebhookConfiguration` and `ApplyValidatingWebhookConfiguration`

sttts commented on pkg/operator/resource/resourceapply/admissions.go:
```diff
@@ -2,76 +2,159 @@ package resourceapply
+	// and ignore clientConfig.caBundle changes on "inject-cabundle" label
+	if required.GetAnnotations() != nil && required.GetAnnotations()[genericCABundleInjectorAnnotation] != "" {
+		copyMutatingWebhookCABundle(existing, required)
+	}
```
can't we check in requiredOriginal whether the CA is set and keep the one on the server if that's not the case? I don't like a dependency on service-ca here. It can be other actors setting the CA.

Danil-Grigorev replied:
I changed it and then checked the operator behavior... It forces the caBundle, so will revert the change: https://github.com/openshift/service-ca-operator/blob/master/pkg/controller/cabundleinjector/mutatingwebhook.go#L48

Danil-Grigorev commented on pkg/operator/resource/resourceapply/admissions.go:
```diff
@@ -2,76 +2,159 @@ package resourceapply
+		return nil, false, fmt.Errorf("Unexpected nil instead of an object")
+	}
+	required := requiredOriginal.DeepCopy()
+
```
This is one thing which I don't think we should do. That mix of hashing and generation is what was causing the necessity of using server-side apply. If you really require to see the full populated spec of the resource, as the server returns it, you need to copy defaulting methods from the server. Otherwise, you have to merge the new spec with the previous one (which we could avoid). In any case, this brings chaos in the approach and make it actually difficult to maintain the state with other operators doing changes on the webhooks with the mentioned annotation.

If you'd like to force redeployment of the resource, the expected generation should be set to 0. This is also replicated on operator startup, as the actual cache for the generation will be empty.

I'll add a test case that will provide an example.

jsafrane replied:
Hash and generation serve different purposes.
* Generation (in `existing`) signalizes that something outside of the operator has modified the WebhookConfiguration and the operator should overwrite these changes.
* Hash of `required` detects changes in `required` - e.g. when a new operator with say new `webhook[].timeout` starts, it should update existing WebhookConfiguration. It does not hash `existing` with defaulted values and caBundle and whatnot! The hash should be very very stable, unless the operator really needs a different `required` than it provided previously. We do the same for Deployments / DaemonSets and they don't mess up the generation at all.

Danil-Grigorev replied:
There are corner cases. Let's say we just created a new webhookConfiguration and didn't use some of the default values which are populated by the server, instead we set the same on the resource ourselves for "better visibility". Then after some time, we remove those - from the server perspective those resources are absolutely equal, so it should not do anything. But it has to apply a new version just because of the hash.

jsafrane replied:
IMO that's expected and OK. Hash is not a perfect solution, it may lead to unnecessary updates. But the update will be just one (to get the hash into annotation), not an endless loop of updates.

Danil-Grigorev replied:
This just seems like an unnecessary complication. Generation covers this use-case too, and we won't have to copy fields with changed certificates from https://github.com/openshift/service-ca-operator to calculate the hash correctly on each reconcile. The code is less error-prone IMO when we rely on a single source of truth, and generation provides a stable one.

jsafrane replied:
No, Generation does not cover this use case at all. And these are not a corner case either.

* Say, you have operator v4.7 that deploys ValidatingWebhookConfiguration with 1 webhook. It gets applied and gets `Generation: 2` and the `operator.Status.Generations[..].LastGeneration: 2`.
* In operator v4.8, you add a new webhook, so the array has now 2 items.
  * `operator.Status` has still `Generations[..].LastGeneration: 2`
  *  ValidatingWebhookConfiguration has `Generation: 2`, which matches the expected one.

-> The current code won't update  the object at all! There will be still only one webhook[]. It will introduce hard to debug issues (trust me, been there...). Ad that's why we added hash of spec to Deployment+DaemonSet and we should add them to *WebhookConfiguration too.

Danil-Grigorev replied:
In this case, this is managed on a higher level. There is surely at least one restart for the operator, so the leader for handling this state will change. The approach is different from the typical one, but it is still the same from a functional perspective. We just store this "hash" approximation in the controller cache, which makes more sense, as the operator's responsibility is to apply his version of the resource at least once, but not more than that. In the case of the stored hash on the resource, it is worse - the new operator may not ever put his version of resource even on startup if someone would sneak in the correct spec hash, but pass a different spec content while the operator was down/upgrading etc..

jsafrane replied:
> as the operator's responsibility is to apply his version of the resource at least once,
Why an operator must apply it at least once? `Apply*WebhookConfig` can see that everything is fine (using this hash) and nothing is needed.

Anyway, I can apply the has in the operator, but I'd bet some users of Apply* will get it wrong.

sttts commented on pkg/operator/resource/resourceapply/admissions.go:
```diff
@@ -2,76 +2,159 @@ package resourceapply
+// and an update performed if the validatingwebhookconfiguration spec and metadata differ from
+// the previously required spec and metadata based on generation change.
+//
+// expectedGeneration set to 0 disables generation tracking, forcing redeployment logic.
```
what is "forcing redeployment logic" ?

sttts replied:
This signature breaks every other user of the func. Do we want that? `ApplyValidatingWebhookConfigurationWithGeneration` would avoid that.

sttts replied:
Let's break consumers IMO. According to @jsafrane there is no other consumer so.

Danil-Grigorev replied:
This is used in this PR in `ApplyDirectly`. Maybe it should not be here in the first place, I'm not sure about a use-case either. But this comment is nothing new, just highlighting the side-effect of the chosen algorithm on Deployment/DaemonSet apply logic too. For some reason, they use `forceRollout`, but it is not needed - https://github.com/openshift/library-go/blob/c8d0c55677d43fef621e8afb5a2c092ed75a5baf/pkg%2Foperator%2Fresource%2Fresourceapply%2Fapps.go#L107

sttts commented on pkg/operator/resource/resourceapply/admissions.go:
```diff
@@ -2,76 +2,159 @@ package resourceapply
+	toWrite.Webhooks = required.Webhooks
 
-	actual, err := client.MutatingWebhookConfigurations().Update(context.TODO(), required, metav1.UpdateOptions{})
+	klog.V(4).Infof("ValidatingWebhookConfiguration %q changes: %v", required.GetNamespace()+"/"+required.GetName(), JSONPatchNoError(existing, toWrite))
```
there was a reason for `if klog.V(4).Enabled() {`

Danil-Grigorev replied:
It should be the same, according to docs: 
```golang
// Thus, one may write either
//	if glog.V(2).Enabled() { klog.Info("log this") }
// or
//	klog.V(2).Info("log this")
// The second form is shorter but the first is cheaper if logging is off because it does
// not evaluate its arguments.
```

sttts commented on pkg/operator/resource/resourceapply/admissions.go:
```diff
@@ -2,76 +2,159 @@ package resourceapply
+}
+
+// copyValidatingWebhookCABundle populates webhooks[].clientConfig.caBundle fields from existing resource
+func copyValidatingWebhookCABundle(existing, required *admissionregistrationv1.ValidatingWebhookConfiguration) {
```
rename params to `from, to`

sttts commented on pkg/operator/resource/resourceapply/admissions.go:
```diff
@@ -2,76 +2,159 @@ package resourceapply
+	toWrite := existingCopy // shallow copy so the code reads easier
+	// Providing upgrade compatibility with service-ca-bundle operator
+	// and ignore clientConfig.caBundle changes on "inject-cabundle" label
+	if required.GetAnnotations() != nil && required.GetAnnotations()[genericCABundleInjectorAnnotation] != "" {
```
I had a previous comment about this which seem to be gone: why don't we check for a CA cert in required, and keep the existing one if there isn't one enforced. I don't like this dependency on service-ca here. There might be different actors creating CA bundles.

Danil-Grigorev replied:
Sorry, I hoped it was unresolved. If we do that an the operator just does it's job, then we will get to a hot loop situation with such an approach - https://github.com/openshift/library-go/pull/836/files#r533587142

Danil-Grigorev commented:
@sttts @jsafrane All updated, only squashed the commits, renamed `admissions.go` to `admissionregistration.go`, and added a small fix to an older [comment](https://github.com/openshift/library-go/pull/836#discussion_r457185342) of yours. PTAL and thanks for all your help reviewing this!

Danil-Grigorev commented:
/retest

sttts commented on pkg/operator/resource/resourceapply/admissionregistration.go:
```diff
@@ -0,0 +1,146 @@
+		existingMutatingWebhooks[webhook.Name] = webhook
+	}
+
+	webhooks := make([]admissionregistrationv1.MutatingWebhook, len(to.Webhooks))
```
why a new map? Just change the existing one.

sttts replied:
```
for i, wh := range to.Webhooks {
  if existing, ok := fromMap[wh.Name]; ok && wh.ClientConfig.CABundle == nil {
    to.Webhooks[i].ClientConfig.CABundle = existing.ClientConfig.CABundle
  }
}

sttts commented:
Up to https://github.com/openshift/library-go/pull/836#discussion_r536091531 lgtm.

@jsafrane can you do a final review? I will approve when lgtm'ed.

jsafrane commented:
/lgtm

sttts commented:
/approve
